pkg/reconciler/revision/reconcile_resources.go

```diff
@@ -66,32 +72,52 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1alpha1.Revi
-	if *deployment.Spec.Replicas > 0 && deployment.Status.AvailableReplicas == 0 {
+	if *deployment.Spec.Replicas > deployment.Status.AvailableReplicas {
 		pods, err := c.KubeClientSet.CoreV1().Pods(ns).List(metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(deployment.Spec.Selector)})
 		if err != nil {
 			logger.Errorf("Error getting pods: %v", err)
 		} else if len(pods.Items) > 0 {
-			// Arbitrarily grab the very first pod, as they all should be crashing
-			pod := pods.Items[0]
-
-			// Update the revision status if pod cannot be scheduled(possibly resource constraints)
-			// If pod cannot be scheduled then we expect the container status to be empty.
-			for _, cond := range pod.Status.Conditions {
-				if cond.Type == corev1.PodScheduled && cond.Status == corev1.ConditionFalse {
-					rev.Status.MarkResourcesUnavailable(cond.Reason, cond.Message)
-					break
+			// Should change revision status if all pods are crashing
+			shouldMarkRev := deployment.Status.AvailableReplicas == 0
+			for _, pod := range pods.Items {
+				// Update the revision status if pod cannot be scheduled(possibly resource constraints)
+				// If pod cannot be scheduled then we expect the container status to be empty.
+				for _, cond := range pod.Status.Conditions {
+					if cond.Type == corev1.PodScheduled && cond.Status == corev1.ConditionFalse {
+						userFacingLogger.Warnf("Pod %s of revision %s cannot be scheduled: %s/%s", pod.Name, rev.Name, cond.Reason, cond.Message)
+						if shouldMarkRev {
+							logger.Infof("%s marking resources unavailable with: %s: %s", rev.Name, cond.Reason, cond.Message)
+							rev.Status.MarkResourcesUnavailable(cond.Reason, cond.Message)
+						}
+						break
+					}
 				}
-			}
 
-			for _, status := range pod.Status.ContainerStatuses {
-				if status.Name == rev.Spec.GetContainer().Name {
-					if t := status.LastTerminationState.Terminated; t != nil {
-						logger.Infof("%s marking exiting with: %d/%s", rev.Name, t.ExitCode, t.Message)
-						rev.Status.MarkContainerExiting(t.ExitCode, t.Message)
-					} else if w := status.State.Waiting; w != nil && hasDeploymentTimedOut(deployment) {
-						logger.Infof("%s marking resources unavailable with: %s: %s", rev.Name, w.Reason, w.Message)
-						rev.Status.MarkResourcesUnavailable(w.Reason, w.Message)
+				for _, status := range pod.Status.ContainerStatuses {
+					// This is based on the fact that rev.Spec.GetContainer() returns
+					// the user container.
+					if status.Name == rev.Spec.GetContainer().Name {
+						if t := status.LastTerminationState.Terminated; t != nil {
+							userFacingLogger.Warnf("Container %s in pod %s of revision %s is in terminated status: %s/%s",
+								status.Name, pod.Name, rev.Name, t.Reason, t.Message)
+							if shouldMarkRev {
+								logger.Infof("%s marking exiting with: %d/%s: %s", rev.Name, t.ExitCode, t.Reason, t.Message)
+								rev.Status.MarkContainerExiting(t.ExitCode, t.Reason, t.Message)
```

Comment:
So if I have 3 pods one is ok, 2 are crashing but for different reasons, you'll set the status twice to different things?